test/ServerTest.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Julian Mundhahs (mundhahj@tf.uni-freiburg.de)
+
+#include <engine/Server.h>
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <boost/beast/http.hpp>
+
+#include "util/GTestHelpers.h"
+#include "util/http/HttpUtils.h"
+#include "util/http/UrlParser.h"
+
+namespace http = boost::beast::http;
+using namespace ad_utility;
+
+TEST(ServerTest, parseHttpRequest) {
+  auto ParsedRequest = [](const std::string& path,
+                          const HashMap<std::string, std::string>& parameters,
+                          const std::optional<std::string>& query)
+      -> testing::Matcher<const UrlParser::ParsedRequest> {
+    return testing::AllOf(
+        AD_FIELD(UrlParser::ParsedRequest, ParsedRequest::path_,
+                 testing::Eq(path)),
+        AD_FIELD(UrlParser::ParsedRequest, ParsedRequest::parameters_,
+                 testing::ContainerEq(parameters)),
+        AD_FIELD(UrlParser::ParsedRequest, ParsedRequest::query_,
+                 testing::Eq(query)));
+  };
+  auto MakeBasicRequest = [](http::verb method, const std::string& target) {
+    return http::request<http::string_body>{method, target, 11};
+  };
+  auto MakeGetRequest = [&MakeBasicRequest](const std::string& target) {
+    return MakeBasicRequest(http::verb::get, target);
+  };
+  auto MakePostRequest = [&MakeBasicRequest](const std::string& target,
+                                             const std::string& contentType,
+                                             const std::string& body) {
+    auto req = MakeBasicRequest(http::verb::post, target);
+    req.set(http::field::content_type, contentType);
+    req.body() = body;
+    req.prepare_payload();
+    return req;
+  };
+  auto parse = [](const httpUtils::HttpRequest auto& request) {
+    return Server::parseHttpRequest(request);
+  };
+  // TODO `default-graph-uri` and `named-graph-uri` can appear multiple times.
+  // This currently throws an exception and is not considered.
```

Comment:
Okay, So this means that in HTTP there can be duplicate parameters in general, and the semantics depend on whatever the service does.
So probably, the value type of the parameters should not be `string` but `HashSet<string>` or something like this.
Then the consuming code (in this case the `Server.cpp` has to check, whiich of these can and cannot appear multiple times. At least add a comment in the respective code in the Parser, s.t. we can revisit it at a later time (I currently am working on the named graph support).

Reply:
The specs (RFC and similar) actually don't specify semantics for multiple values for a query parameter.
What would be the reason against simply using a `multiset` instead of a `set` here? Is it that abseil does not provide a multiset and we'd have to resort to `std::multiset`?
This comment is just a temporary note for me. For now I'd add some TODOs in appropriate places and warnings to the result if these parameters are used.